File: planner/app/src/main/java/com/sunilson/firenote/activities/BinActivity.java
package com.sunilson.firenote.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.OvershootInterpolator;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.sunilson.firenote.BaseApplication;
import com.sunilson.firenote.Interfaces.BinInterface;
import com.sunilson.firenote.Interfaces.ConfirmDialogResult;
import com.sunilson.firenote.Interfaces.MainActivityInterface;
import com.sunilson.firenote.ItemTouchHelper.SimpleItemTouchHelperCallbackMain;
import com.sunilson.firenote.R;
import com.sunilson.firenote.adapters.BinRecyclerAdapter;
import com.sunilson.firenote.baseClasses.Element;
import com.sunilson.firenote.dialogs.ConfirmDialog;

import jp.wasabeef.recyclerview.adapters.AlphaInAnimationAdapter;
import jp.wasabeef.recyclerview.animators.ScaleInAnimator;

public class BinActivity extends BaseActivity implements BinInterface, ConfirmDialogResult {

    private DatabaseReference mReference, mBinReference, mElementsRefernce, mContentsReference;
    private FirebaseUser user;
    private RecyclerView binList;
    private boolean restore = false;
    private View.OnClickListener recycleOnClickListener;
    private View.OnLongClickListener recycleOnLongClickListener;
    private ChildEventListener mBinListener;
    private BinRecyclerAdapter binRecyclerAdapter;
    private CoordinatorLayout coordinatorLayout;
    private Element currentlySelectedElement;
    private String elementID, elementName;
    private MainActivityInterface mainActivityInterface;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bin);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        mainActivityInterface = (MainActivityInterface) ((BaseApplication) getApplicationContext()).mainContext;
        user = mAuth.getCurrentUser();

        //Get passed arguments
        Intent i = getIntent();
        elementID = i.getStringExtra("elementID");
        elementName = i.getStringExtra("elementName");

        //If logged in
        if (user != null) {
            mReference = FirebaseDatabase.getInstance().getReference().child("users").child(user.getUid());
            coordinatorLayout = (CoordinatorLayout) findViewById(R.id.activity_bin);

            //Initialize List
            binList = (RecyclerView) findViewById(R.id.binList);
            binList.setHasFixedSize(true);
            LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
            binList.setLayoutManager(linearLayoutManager);
            binRecyclerAdapter = new BinRecyclerAdapter(this, recycleOnClickListener, recycleOnLongClickListener);
            AlphaInAnimationAdapter alphaInAnimationAdapter = new AlphaInAnimationAdapter(binRecyclerAdapter);
            alphaInAnimationAdapter.setFirstOnly(false);
            alphaInAnimationAdapter.setDuration(200);
            binList.setAdapter(alphaInAnimationAdapter);
            binList.setItemAnimator(new ScaleInAnimator(new OvershootInterpolator(1f)));
            binList.getItemAnimator().setAddDuration(400);
            ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallbackMain(binRecyclerAdapter);
            ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);
            itemTouchHelper.attachToRecyclerView(binList);
            initializeBinListener();

            //Click listeners
            initializeOnClickListener();
            initializeOnLongClickListener();

            //Get reference (bundle or not)
            if (elementID == null) {
                mElementsRefernce = mReference.child("elements").child("main");
                mBinReference = mReference.child("bin").child("main");
                setTitle(getString(R.string.bin));
            } else {
                mElementsRefernce = mReference.child("elements").child("bundles").child(elementID);
                mBinReference = mReference.child("bin").child("bundles").child(elementID);
                setTitle(getString(R.string.bin) + " - " + elementName);
            }

            mContentsReference = mReference.child("contents");
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        if (mReference != null) {
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    mBinReference.addChildEventListener(mBinListener);
                }
            }, 200);
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        if (mBinListener != null) {
            mBinReference.removeEventListener(mBinListener);
        }

        binRecyclerAdapter.clear();
    }

    /**
     * Listener for bin elements
     */
    private void initializeBinListener() {
        mBinListener = new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                Element element = dataSnapshot.getValue(Element.class);
                element.setElementID(dataSnapshot.getKey());
                binRecyclerAdapter.add(element);
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
                Element element = dataSnapshot.getValue(Element.class);
                binRecyclerAdapter.remove(element.getElementID());

                //If element was removed by a restore request
                if (restore) {
                    //Move element to "elements" and display "restored"-message
                    element.setElementID(element.getElementID());
                    mElementsRefernce.child(element.getElementID()).setValue(element);
                    Snackbar snackbar = Snackbar
                            .make(coordinatorLayout, R.string.element_restored, 6000);
                    snackbar.show();
                    restore = false;
                } else {
                    //If element was deleted from bin, also delete its contents
                    if (element.getNoteType().equals("bundle")) {
                        mReference.child("elements").child("bundles").child(element.getElementID()).removeValue();
                    } else {
                        mContentsReference.child(element.getElementID()).removeValue();
                    }
                }
            }


            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        };
    }

    /**
     * Display Restore dialog on long click
     */
    private void initializeOnLongClickListener() {
        recycleOnLongClickListener = new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (!restore) {
                    int itemPosition = binList.getChildLayoutPosition(v);
                    currentlySelectedElement = binRecyclerAdapter.getItem(itemPosition);
                    DialogFragment dialogFragment = ConfirmDialog.newInstance(getString(R.string.restore_element), getString(R.string.restore_element_question), "restore", null);
                    dialogFragment.show(getSupportFragmentManager(), "dialog");
                }

                return true;
            }
        };
    }

    /**
     * Dislay restore dialog on click
     */
    private void initializeOnClickListener() {
        recycleOnClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!restore) {
                    int itemPosition = binList.getChildLayoutPosition(v);
                    currentlySelectedElement = binRecyclerAdapter.getItem(itemPosition);
                    DialogFragment dialogFragment = ConfirmDialog.newInstance(getString(R.string.restore_element), getString(R.string.restore_element_question), "restore", null);
                    dialogFragment.show(getSupportFragmentManager(), "dialog");
                }
            }
        };
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();

        if (id == android.R.id.home) {
            this.finish();
            return true;
        } else if (id == R.id.bin_delete_all) {
            initializeDeleteAllDialog();
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * Create wipe confirm dialog
     */
    private void initializeDeleteAllDialog() {
        DialogFragment dialogFragment = ConfirmDialog.newInstance(getString(R.string.clear_bin_title), getString(R.string.clear_bin_question), "clear", null);
        dialogFragment.show(getSupportFragmentManager(), "dialog");
    }

    /**
     * Get results from confirm dialog
     *
     * @param result
     * @param type
     * @param args
     */
    @Override
    public void confirmDialogResult(boolean result, String type, Bundle args) {
        if (result) {
            if (type.equals("clear")) {
                //Remove all entries in the bin
                mBinReference.removeValue();
            } else if (type.equals("restore")) {
                //Restore single element of bin
                if (currentlySelectedElement != null) {
                    mBinReference.child(currentlySelectedElement.getElementID()).removeValue();
                    if(!((BaseApplication) getApplicationContext()).getInternetConnected()) {
                        Toast.makeText(this, R.string.edit_no_connection, Toast.LENGTH_LONG).show();
                    }
                }
                restore = true;
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_bin, menu);
        return true;
    }

    /**
     * Get reference to all bin elements
     *
     * @return All elements that are currently in bin
     */
    @Override
    public DatabaseReference getBinReference() {
        return mBinReference;
    }
}

File: planner/app/src/main/java/com/sunilson/firenote/presentation/bin/BinActivity.kt
package com.sunilson.firenote.presentation.bin

import android.content.Context
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.ActionBar
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.OvershootInterpolator
import com.sunilson.firenote.R
import com.sunilson.firenote.presentation.elements.elementList.ElementRecyclerAdapter
import com.sunilson.firenote.presentation.elements.elementList.ElementRecyclerAdapterFactory
import com.sunilson.firenote.presentation.shared.base.BaseActivity
import com.sunilson.firenote.presentation.shared.changeStatusBarColor
import com.sunilson.firenote.presentation.shared.dialogs.ConfirmDialog
import com.sunilson.firenote.presentation.shared.dialogs.interfaces.DialogListener
import com.sunilson.firenote.presentation.shared.interfaces.HasElementList
import jp.wasabeef.recyclerview.animators.ScaleInAnimator
import kotlinx.android.synthetic.main.activity_bin.*
import javax.inject.Inject

class BinActivity : BaseActivity(), BinPresenterContract.View, HasElementList {

    @Inject
    lateinit var binPresenter: BinPresenter

    @Inject
    lateinit var elementRecyclerAdapterFactory: ElementRecyclerAdapterFactory

    private var elementName: String? = null
    private var elementColor: Int? = null
    override lateinit var adapter: ElementRecyclerAdapter
    override var parent: String? = null

    override val mContext: Context
        get() = this

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_bin)
        setSupportActionBar(findViewById<View>(R.id.toolbar) as Toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        parent = intent.getStringExtra("elementID")
        elementName = intent.getStringExtra("elementName")
        elementColor = intent.getIntExtra("elementColor", -1)

        elementColor?.let {
            if (it != -1) {
                window.changeStatusBarColor(elementColor as Int)
                toolbar.setBackgroundColor(elementColor as Int)
                toolbar.setTitleTextColor(resources.getColor(R.color.white_text_color))
                val backButton = ContextCompat.getDrawable(this, R.drawable.abc_ic_ab_back_material)
                backButton?.setTint(resources.getColor(R.color.white_text_color))
                (supportActionBar as ActionBar).setHomeAsUpIndicator(backButton)
            }
        }

        if (elementName != null) title = elementName + " " + getString(R.string.bin)

        //Initialize list
        binList.setHasFixedSize(true)
        binList.itemAnimator = ScaleInAnimator(OvershootInterpolator(1f))
        binList.itemAnimator.addDuration = 300
        binList.layoutManager = LinearLayoutManager(this)
        adapter = elementRecyclerAdapterFactory.create(View.OnClickListener { view ->
            val dialog = ConfirmDialog.newInstance(getString(R.string.restore_element), getString(R.string.restore_element_question))
            dialog.listener = object : DialogListener<Boolean> {
                override fun onResult(result: Boolean?) {
                    if (result == true) {
                        binPresenter.restoreElement(adapter.data[binList.getChildLayoutPosition(view)].elementID)
                    }
                }
            }
            dialog.show(supportFragmentManager, "dialog")
        }, View.OnLongClickListener { _ -> true }, { id, _ ->
            binPresenter.deleteElement(id)
        }, binList)
        adapter.swipeable = false
        binList.adapter = adapter
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_bin, menu)

        if (elementColor != null && elementColor != -1) {
            val drawable = menu?.findItem(R.id.bin_delete_all)?.icon
            drawable?.setTint(resources.getColor(R.color.white_text_color))
            menu?.findItem(R.id.bin_delete_all)?.icon = drawable
        }

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            android.R.id.home -> {
                finish()
            }
            R.id.bin_delete_all -> {
                val dialog = ConfirmDialog.newInstance(getString(R.string.clear_bin_title), getString(R.string.clear_bin_question))
                dialog.listener = object : DialogListener<Boolean> {
                    override fun onResult(result: Boolean?) {
                        if (result == true) {
                            binPresenter.clearElements()
                        }
                    }
                }
                dialog.show(supportFragmentManager, "dialog")
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun toggleLoading(loading: Boolean, message: String?) {}
}